proxy/src/transparency/tcp.rs

```diff
@@ -71,14 +73,177 @@ impl Proxy {
         let fut = connect.connect()
             .map_err(|e| debug!("tcp connect error: {:?}", e))
             .and_then(move |tcp_out| {
-                let (in_r, in_w) = tcp_in.split();
-                let (out_r, out_w) = tcp_out.split();
-
-                copy(in_r, out_w)
-                    .join(copy(out_r, in_w))
-                    .map(|_| ())
+                Duplex::new(tcp_in, tcp_out)
                     .map_err(|e| debug!("tcp error: {}", e))
             });
         Box::new(fut)
     }
 }
+
+/// A future piping data bi-directionally to In and Out.
+struct Duplex<In, Out> {
+    in_buf: CopyBuf,
+    in_eof: bool,
+    in_io: In,
+    in_shutdown: bool,
```

Comment:
The repetition here makes me thing that we should have this be something like:

```rust
struct Duplex<In, Out> {
    in: EndPoint<In>,
    out: EndPOint<Out>,
}

struct EndPoint { // Probably there is a better name for this. 
    buf: CopyBuf,
    eof: bool,
    io: In,
    shutdown: bool,
}
```

Also, are some of these variables mutually exclusive with each other; i.e. should some part of this be an enum that helps enforce a well-typed state machine?